src/core/server/saved_objects/migrations/types.ts

```diff
@@ -17,19 +17,14 @@
  * under the License.
  */
 
-import { SavedObjectsSchema } from './schema';
+import { RawSavedObjectDoc } from '../serialization';
+import { SavedObjectsMigrationLogger } from './core/migration_logger';
 
-type Schema = PublicMethodsOf<SavedObjectsSchema>;
-const createSchemaMock = () => {
-  const mocked: jest.Mocked<Schema> = {
-    getIndexForType: jest.fn().mockReturnValue('.kibana-test'),
-    isHiddenType: jest.fn().mockReturnValue(false),
-    isNamespaceAgnostic: jest.fn((type: string) => type === 'global'),
-    getConvertToAliasScript: jest.fn().mockReturnValue(undefined),
-  };
-  return mocked;
-};
+export type SavedObjectMigrationFn = (
+  doc: RawSavedObjectDoc,
+  log: SavedObjectsMigrationLogger
```

Comment:
> Would there be other runtime params?

Not atm. In migration v2, maybe?

> We can make it extensible for further changes. For example:

This would be way more futur-proof, however:
- This would breaks 'legacy' registered migrations that are expecting `log` as the concrete second parameter (I could wrap the legacy migration handlers to address this, but it adds a layer of complexity for something not used atm)

- (More importantly) I think we shouldn't alter migrations too much outside of the scope of 'SO migration v2'. I also have the feeling the v2 migration will impact this part of SO more deeply than than (I see the 'async' keyword arriving in the signature for v2....)

@rudolf What do you think? Would this change be of help for the futur of SO migrations? Do you want me to it in current PR?

Reply:
Similar to the `SavedObjectsMigrationLogger` we could make breaking changes and refactor the existing migrations, but since many migrations are written in javascript and a single uncaught exception can prevent kibana from starting up, this is risky. So I think it's better we keep the API and refactor it post 8.0.0.

Reply:
For the logger, we are forced to wait until everything is migrated to TS to be sure we don't break anything.

For this `SavedObjectMigrationFn` signature, it would be possible to wrap the legacy migration methods to call them with only `context.logger` as the second parameter (as I'm already wrapping/converting the uiExport SO-related things atm), so this can be done in current PR without the risk to break something.

Reply:
After going through all the existing migrations listed here https://github.com/elastic/kibana/issues/49764#issuecomment-563902522 only a few migrations use the logger, and they only use the logger to print a warning when they ignore an exception. When we introduce the invalid object tagging https://github.com/elastic/kibana/issues/55406 this will become a big anti-pattern. Migrations should either succeed or throw and we might want to "fix" these existing migrations. When they throw core will log the error and the stack, but the migration itself doesn't need to do any logging.

In addition, migrations in NP have access to the scope of the `setup` function, so if they needed to access a logger or for instance a configuration option, they wouldn't need this to be passed in via the context.

So long term I imagine a migration function will just be a transform function `SavedObjectMigrationFn = (doc: RawSavedObjectDoc) => RawSavedObjectDoc`.

> it would be possible to wrap the legacy migration methods

If we do this we should also expose a utility for plugins to be able to wrap their legacy migrations when they port them to NP. But I think it might be simpler to just mark the second parameter (`SavedObjectsMigrationLogger`) as deprecated.